### PR TITLE
Issue #7649: Update doc for ModifiedControlVariable

### PR DESCRIPTION
**Pull request by SGanguly1999**
Closes: #7649

image

`$ cat config.xml`

```
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <property name="tabWidth" value="7"/>
    <module name="ModifiedControlVariable"/>

  </module>
</module>
```

`$ cat Demo.java`

```
class Demo {
    public static void main(String args[])
    {
        for(int i=0;i<8;i++)
        {
            i++; // violation
        }

    }
}
```
Output:-
```
$ java -jar checkstyle-8.39-all.jar -c config.xml Demo.java
Starting audit...
[ERROR] E:\New folder\Java Dev\src\com\company\Demo.java:8:14: Control variable 'i' is modified. [ModifiedControlVariable]
Audit done.
Checkstyle ends with 1 errors.

```

image

`$ cat config.xml`

```
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <property name="tabWidth" value="7"/>
    <module name="ModifiedControlVariable">
      <property name="skipEnhancedForLoopVariable" value="true"/>
    </module>
  </module>
</module>
```
`$ cat Demo.java`

```
import java.util.*;
class Demo
{
public static void main(String args[])
{
List<String> lines =new ArrayList<String>();
lines.add(" CHECKSTYLE ");
for (String line: lines) {
line = line.trim();   // it will skip this violation
}
}
}
```

Output:-

```
$ java -jar checkstyle-8.39-all.jar -c config.xml Demo.java
Starting audit...
Audit done.
```

### CODE REVIEW

**Comment by romani** [comment]
Github, generate web site.

**Comment by SGanguly1999** [comment]
Github, generate website

**Comment by SGanguly1999** [comment]
@romani I updated the doc by specifing the loop structure.But don't know why this Cirrus-JDK15 is felling

**Comment by SGanguly1999** [comment]
Github, generate website

**Comment by romani** [comment]
Please fix formatting in your examples https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/controlvariable_2021165225/config_coding.html#ModifiedControlVariable

**Comment by SGanguly1999** [comment]
> Please fix formatting in your examples https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/controlvariable_2021165225/config_coding.html#ModifiedControlVariable

@romani could you kindly tell me which portion i need to make changes(Is that the alignment of braces?)

**Comment by romani** [comment]
Yes, take a look at other examples they are good for reading, indetantion is good for all lines.

**Comment by SGanguly1999** [comment]
> Yes, take a look at other examples they are good for reading, indetantion is good for all line
Okay...sure

File: src/xdocs/config_coding.xml [hunk 1]
```diff
@@ -4335,6 +4335,16 @@ for (int i = 0; i &lt; 1; i++) {
...
+        </source>
+        <p>
+          NOTE:The check works with only primitive type variables.
+          The check will not work for arrays used as control variable.An example is
+        </p>
+        <source>
+for (int a[]={0};a[0] &lt; 10;a[0]++)
+{
+a[0]++;   // it will skip this violation
 }
```
Comment (aryaniiit002) [inline src/xdocs/config_coding.xml:4348]: fix your indentation then please add `// OK` or `// violation` keyword before its description.

File: src/xdocs/config_coding.xml [hunk 2]
```diff
@@ -4396,8 +4420,17 @@ for (int i = 0; i &lt; 10;) {
...
+public static void main(String args[])
+{
+List&lt;String&gt; lines =new ArrayList&lt;String&gt;();
+lines.add(" CHECKSTYLE ");
 for (String line: lines) {
-  line = line.trim();   // it will skip this violation
+line = line.trim();   // it will skip this violation
+}
+}
 }
```
Comment (aryaniiit002) [inline src/xdocs/config_coding.xml:4434]: same

File: src/xdocs/config_coding.xml [hunk 3]
```diff
@@ -4376,7 +4386,21 @@ for (int i = 0; i &lt; 10;) {
...
+          Example:
+        </p>
+        <source>
+class Demo
+{
+public static void main(String args[])
+{
+for(int i=0;i &lt; 8;i++)
+{
+i++; // violation
```
Comment (aryaniiit002) [inline src/xdocs/config_coding.xml:4399]: can you describe this violation a little bit in comments after `// violation`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java [hunk 4]
```diff
@@ -57,6 +57,15 @@
  *   i++;
  * }
  * </pre>
+ * <p>
+ * NOTE:The check works with only primitive type variables.
+ * The check will not work for arrays used as control variable.An example is
+ * </p>
+ * <pre>
+ * for (int a[]={0};a[0] &lt; 10;a[0]++) {
+ *  a[0]++;   // it will skip this violation
```
Comment (aryaniiit002) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:66]: so many spaces btw `;` and `//`

**Comment by aryaniiit002** [comment]
Please have a look at this example https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/controlvariable_2021165225/config_coding.html#MissingCtor_Examples

**Comment by SGanguly1999** [comment]
@aryaniiit002 could you tell me the exact reason of the failure?

**Comment by SGanguly1999** [comment]
image

I tried to follow this format.
Initially gave one spaces but the code does'nt look good enough to be read so I added such spacing.Infact in this example it is applied too .

**Comment by SGanguly1999** [comment]
@romani is the formatting alright?

**Comment by SGanguly1999** [comment]
Github, generate website

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java [hunk 5]
```diff
@@ -92,8 +113,15 @@
...
  *
  * <pre>
- * for (String line: lines) {
- *   line = line.trim();   // it will skip this violation
+ * import java.util.*;
+ * class Demo {
+ *    public static void main(String args[]) {
+ *      List&lt;String&gt; lines =new ArrayList&lt;String&gt;();
+ *      lines.add(" CHECKSTYLE ");
+ *      for (String line: lines) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:121]: this is overkill, please read
https://www.geeksforgeeks.org/for-each-loop-in-java/

please use 
```
for (String arg: args) {
  arg = arg.trim();   // ok
}
```
no imports are required.
please add this example to code for default config to show that violation will be on that line.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java (hunk 5 above)
Comment (SGanguly1999) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:121]: @romani  sure I will replace it. But is there any mistake in my example?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:121]: No mistake, just keep it simple and avoid extra code

**Comment by SGanguly1999** [comment]
@romani made the changes

**Comment by SGanguly1999** [comment]
@romani please see whether the changes asked by you correctly done by me or not

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java [hunk 6]
```diff
@@ -92,8 +113,8 @@
  * <p>Example:</p>
  *
  * <pre>
- * for (String line: lines) {
- *   line = line.trim();   // it will skip this violation
+ * for (String arg: args) {
+ *   arg = arg.trim();   // ok
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:117]: Please move this forech loop to first example to show that there is no violation on such code by default.
Please use absolutely the same java code for each  example

**Review by romani — CHANGES_REQUESTED** [review]
Item:

**Comment by SGanguly1999** [comment]
@romani made the changes. Please review whether everything is alright or not

**Comment by romani** [comment]
Github, generate web site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java [hunk 7]
```diff
@@ -73,6 +82,18 @@
...
  * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * for(int i=0;i &lt; 8;i++) {
+ *   i++; // violation, control variable modified
+ * }
+ * String args1[]={"Coding", "block"};
+ * for (String arg: args1) {
+ *   arg = arg.trim(); // violation,control variable modified
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: add space after `,`

**Review by romani — CHANGES_REQUESTED** [review]
last items:

**Comment by SGanguly1999** [comment]
@romani made the changes

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java (hunk 7 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: not done.
please fix it and reply ''done" to this thread after you see that changes are visible in github

Comment (SGanguly1999) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: image

I added space after "Coding" in args1[ ].@romani is that you are asking for?

Comment (SGanguly1999) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: Github is not showing this one outdated

Comment (SGanguly1999) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: Extremely sorry,I  added it after "Coding"

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java:93]: `violation,control variable modified` ==> `violation, control variable modified`

**Comment by SGanguly1999** [comment]
Github, generate website.

**Comment by SGanguly1999** [comment]
image
@romani done.

**Comment by SGanguly1999** [comment]
image
@romani done

**Review by romani — APPROVED** [review]
ok to merge

**Pull request merged**